### PR TITLE
Extracting property values to make ARM output variables accessible

### PR DESCRIPTION
**Pull request by tariq1890**
**What this PR does / why we need it**: This PR extracts values in code which are normally using template params and ARM variables. This is so done for easier programmatic access to the ARM output params

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: Working towards #3850 

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/acsengine/template_generator.go [hunk 1]
```diff
@@ -802,12 +802,24 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 			imageRef := cs.Properties.MasterProfile.ImageRef
 			return imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
 		},
+		"GetRouteTableName": func() string {
+			return cs.Properties.GetRouteTableName()
```
[the remark this answers is not in the document]
Comment (tariq1890) [reply pkg/acsengine/template_generator.go]: I think it should be tied to the Properties struct. What are thoughts on moving cs.Properties to a variable?

**Comment by tariq1890** [comment]
@jackfrancis @CecileRobertMichon ready for review.

File: pkg/api/types.go [hunk 2]
```diff
@@ -715,6 +719,113 @@ func (p *Properties) HasVirtualMachineScaleSets() bool {
 	return false
 }
 
+// K8sOrchestratorName returns the 3 character orchestrator code for kubernetes-based clusters.
+func (p *Properties) K8sOrchestratorName() string {
+	if p.OrchestratorProfile.IsKubernetes() ||
+		p.OrchestratorProfile.IsOpenShift() {
+		if p.HostedMasterProfile != nil {
+			return "aks"
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: this should be a const as well

is it a new value introduced?

Comment (CecileRobertMichon) [reply pkg/api/types.go]: also should `(p.OrchestratorProfile.IsOpenShift() && p.HostedMasterProfile != nil)` return `aks`?

Comment (CecileRobertMichon) [reply pkg/api/types.go]: also should `(p.OrchestratorProfile.IsOpenShift() && p.HostedMasterProfile != nil)` return `aks`?

Comment (tariq1890) [reply pkg/api/types.go]: no. You'll find this is params_k8s.go

File: pkg/api/types.go [hunk 3]
```diff
@@ -715,6 +719,113 @@ func (p *Properties) HasVirtualMachineScaleSets() bool {
 	return false
 }
 
+// K8sOrchestratorName returns the 3 character orchestrator code for kubernetes-based clusters.
+func (p *Properties) K8sOrchestratorName() string {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: unit test for this new function?

Comment (tariq1890) [reply pkg/api/types.go]: Sure. Missed that one

File: pkg/api/const.go [hunk 4]
```diff
@@ -72,6 +76,8 @@ const (
 	// DefaultKubernetesFirstConsecutiveStaticIPOffsetVMSS specifies the IP address offset of master 0 in VMSS
 	// when VNET integration is enabled.
 	DefaultKubernetesFirstConsecutiveStaticIPOffsetVMSS = 4
+	// DefaultSubnetNameResourceSegmentIndex specifies the default subnet name resource segment index.
+	DefaultSubnetNameResourceSegmentIndex = 10
```
Comment (CecileRobertMichon) [inline pkg/api/const.go]: where is this used?

Comment (tariq1890) [reply pkg/api/const.go]: https://github.com/tariq1890/acs-engine/blob/extract_output_params/pkg/api/types.go#L784

Comment (CecileRobertMichon) [reply pkg/api/const.go]: ah my search failed me

File: pkg/api/types.go [hunk 5]
```diff
@@ -717,6 +721,113 @@ func (p *Properties) HasVMSSAgentPool() bool {
...
+	return p.K8sOrchestratorName() + "-" + p.AgentPoolProfiles[0].Name + "-" + p.GenerateClusterID() + "-vmss"
+}
+
+// IsHostedMasterProfile returns true if the cluster has a hosted master
+func (p *Properties) IsHostedMasterProfile() bool {
+	return p.HostedMasterProfile != nil
+}
+
+// GetSubnetName returns the subnet name of the cluster based on its current configuration.
+func (p *Properties) GetSubnetName() string {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: I don't see most of these functions being used in the var file yet, is it intentional?

Comment (tariq1890) [reply pkg/api/types.go]: yes. This is intended to be non-breaking change. If everyone is okay with what we have so far. We can start using these methods to extract the output parameters in the templates.

**Comment by CecileRobertMichon** [comment]
I really like this 💯

File: pkg/api/types.go (hunk 2 above)
Comment (CecileRobertMichon) [reply pkg/api/types.go]: > also should (p.OrchestratorProfile.IsOpenShift() && p.HostedMasterProfile != nil) return aks?
This is the current behavior so we should leave it that way

File: pkg/api/types.go [hunk 6]
```diff
@@ -65,6 +69,17 @@ type Properties struct {
 	AzProfile               *AzProfile               `json:"azProfile,omitempty"`
 }
 
+// ClusterMetadata represents the metadata of the ACS cluster.
```
Comment (jackfrancis) [inline pkg/api/types.go]: s/ACS/AKS

Comment (jackfrancis) [reply pkg/api/types.go]: or, if it's not AKS-specific (i.e., if it also applies to acs-engine built clusters), then just say "Kubernetes cluster" in the comment

**Comment by jackfrancis** [comment]
this /lgtm with a minor comment

File: pkg/acsengine/defaults.go [hunk 7]
```diff
@@ -720,7 +720,7 @@ func setHostedMasterProfileDefaults(a *api.Properties) {
 
 func setDefaultCerts(a *api.Properties) (bool, error) {
 	if a.MasterProfile != nil && a.OrchestratorProfile.OrchestratorType == api.OpenShift {
-		return certgen.OpenShiftSetDefaultCerts(a, DefaultOpenshiftOrchestratorName, GenerateClusterID(a))
+		return certgen.OpenShiftSetDefaultCerts(a, api.DefaultOpenshiftOrchestratorName, a.GenerateClusterID())
```
Comment (weinong) [inline pkg/acsengine/defaults.go]: nit: `GenerateClusterID` seems to suggest to generate every time it is called... `GetClusterID` seems better?

Comment (tariq1890) [reply pkg/acsengine/defaults.go]: Certainly!

Comment (weinong) [reply pkg/acsengine/defaults.go]: ok.. maybe it's generated after all...

Comment (jackfrancis) [reply pkg/acsengine/defaults.go]: Just to clarify, the existing `GenerateClusterID` (before this change) also generates w/ every invocation as well, right?

Comment (weinong) [reply pkg/acsengine/defaults.go]: Talked to @tariq1890 , it's seeded using the hash. so clusterID is always the same regardless how many times it is called

Comment (weinong) [reply pkg/acsengine/defaults.go]: In this case, I'd like to get `GetClusterID` :)

Comment (tariq1890) [reply pkg/acsengine/defaults.go]: @jackfrancis no, I've added a new field in Properties to store the ClusterID after it's generated during the first time.

**Pull request merged**